File: 双指针/1.反转字符串中的单词/cpp/main.cpp
#include "bits/stdc++.h"
using namespace std;
// https://leetcode.cn/problems/reverse-words-in-a-string/description/
class Solution {
public:
  string reverseWords(string s) {
    int left = s.length() - 1;
    int right = s.length() - 1;
    string ans = "";
    while (left >= 0) {
      char ch = s[left];
      if (ch == ' ') {
        if (left != right) {
          ans += s.substr(left + 1, right - left);
          ans += " ";
          right = left;
        }
        right--;
      }
      left--;
    }
    if (right >= 0)
      ans += s.substr(left + 1, right - left);
    else
      ans = ans.substr(0, ans.size() - 1);
    return ans;
  }
};
int main() {
  Solution s;
  s.reverseWords("  hello world  ");
  return 0;
}

File: 双指针/1.反转字符串中的单词/python/main.py
from functools import cmp_to_key
from typing import List


# import sys
#
# sys.setrecursionlimit(100000)  # 例如这里设置为十万


class Solution:
    def reverseWords(self, s: str) -> str:
        ans = ""
        left = len(s) - 1
        right = len(s) - 1
        while left >= 0:
            ch = s[left]
            if ch == " ":
                if left != right:
                    ans += s[left + 1:right + 1]
                    ans += " "
                    right = left
                right -= 1
            left -= 1
        if right >= 0:
            ans += s[left + 1:right + 1]
        else:
            ans = ans[0:len(ans)-1]
        return ans


if __name__ == '__main__':
    s = Solution()
    s.reverseWords("  hello world  ")
